File: it-agents/skills/code-architect/references/code-examples/ruby/abstract_factory.rb
# frozen_string_literal: true

# Abstract Factory — create families of related objects without specifying concrete classes.

# Product families: a "header" and a "field" that must visually match.

class LightHeader
  def render = "=== light header ==="
end

class LightField
  def render(name) = "[ #{name} ]"
end

class DarkHeader
  def render = "*** dark header ***"
end

class DarkField
  def render(name) = "< #{name} >"
end

# Abstract factory contract: build_header + build_field.
class LightKit
  def build_header = LightHeader.new
  def build_field  = LightField.new
end

class DarkKit
  def build_header = DarkHeader.new
  def build_field  = DarkField.new
end

# Client works only against the factory and product roles.
class Form
  def initialize(kit) = @kit = kit

  def draw(fields)
    lines = [@kit.build_header.render]
    field = @kit.build_field
    fields.each { |name| lines << field.render(name) }
    lines.join("\n")
  end
end

if __FILE__ == $PROGRAM_NAME
  [LightKit.new, DarkKit.new].each do |kit|
    puts Form.new(kit).draw(%w[name email])
    puts
  end
end

File: it-agents/skills/code-architect/references/code-examples/go/abstract_factory.go
// Abstract Factory: create families of related objects without naming
// their concrete types, keeping a whole family mutually consistent.
package main

import "fmt"

// Two product roles that belong to the same family.
type Header interface{ Render() string }
type Footer interface{ Render() string }

type lightHeader struct{}

func (lightHeader) Render() string { return "light header" }

type lightFooter struct{}

func (lightFooter) Render() string { return "light footer" }

type darkHeader struct{}

func (darkHeader) Render() string { return "dark header" }

type darkFooter struct{}

func (darkFooter) Render() string { return "dark footer" }

// ThemeFactory produces a coherent family of widgets.
type ThemeFactory interface {
	NewHeader() Header
	NewFooter() Footer
}

type lightTheme struct{}

func (lightTheme) NewHeader() Header { return lightHeader{} }
func (lightTheme) NewFooter() Footer { return lightFooter{} }

type darkTheme struct{}

func (darkTheme) NewHeader() Header { return darkHeader{} }
func (darkTheme) NewFooter() Footer { return darkFooter{} }

func assemble(f ThemeFactory) {
	fmt.Println(f.NewHeader().Render(), "/", f.NewFooter().Render())
}

func main() {
	assemble(lightTheme{})
	assemble(darkTheme{})
}
